packages/dev/storybook-builder-parcel/index.js

```diff
@@ -0,0 +1,239 @@
+const {Parcel} = require('@parcel/core');
+const path = require('path');
+const {createProxyMiddleware} = require('http-proxy-middleware');
+const {normalizeStories, loadPreviewOrConfigFile} = require('@storybook/core-common');
+const fs = require('fs');
+
+exports.start = async function ({options, router}) {
+  let parcel = await createParcel(options, true);
+
+  router.use(async (req, res, next) => {
+    if (req.url === '/' || req.url === '/index.html') {
+      return next();
+    }
+
+    let proxy = createProxyMiddleware({
+      target: 'http://localhost:3000/',
+      selfHandleResponse: true,
+      logLevel: 'warn',
+      onProxyRes(proxyRes, req, res) {
+        // Parcel dev server responds with main HTML page if the file doesn't exist...
+        if (proxyRes.statusCode === 404 || (proxyRes.headers['content-type']?.startsWith('text/html') && !req.url.startsWith('/iframe.html'))) {
+          return next();
+        } else {
+          res.statusCode = proxyRes.statusCode;
+          for (let header in proxyRes.headers) {
+            res.setHeader(header, proxyRes.headers[header]);
+          }
+          proxyRes.pipe(res);
+        }
+      }
+    });
+
+    // Remove socket/connection temporarily to prevent proxy from subscribing to `close` event and triggering warning.
+    let {socket, connection} = req;
+    req.socket = null;
+    req.connection = null;
+    await proxy(req, res, next);
+    req.socket = socket;
+    req.connection = connection;
+  });
+
+  let subscription = await parcel.watch();
+  process.on('SIGINT', async () => {
+    await subscription.unsubscribe();
+    process.exit();
+  });
+
+  return {
+    async bail(e) {
+      await subscription.unsubscribe();
+    },
+    stats: {},
+    totalTime: 0
+  };
+};
+
+exports.build = async function ({options}) {
+  let parcel = await createParcel(options);
+  await parcel.run();
+};
+
+exports.corePresets = [];
+exports.previewPresets = [];
+
+async function createParcel(options, isDev = false) {
+  await generateHTML(options);
+  await generateJS(options);
+
+  return new Parcel({
+    entries: path.join(__dirname, 'iframe.html'),
+    config: path.resolve(__dirname, '..', '..', '..', '.parcelrc-storybook'),
+    mode: isDev ? 'development' : 'production',
+    serveOptions: isDev ? {
+      port: 3000
+    } : null,
+    hmrOptions: isDev ? {
+      port: 3001
+    } : null,
+    additionalReporters: [
+      {
+        packageName: '@parcel/reporter-cli',
+        resolveFrom: __filename
+      }
+    ],
+    defaultTargetOptions: {
+      distDir: options.outputDir,
+      publicUrl: './',
+      shouldScopeHoist: false // TODO
```

Comment:
todo?

Reply:
I need to upgrade parcel to fix something, but will do that separately.